pkg/netstore/netstore.go

```diff
@@ -33,13 +33,13 @@ func (s *store) Get(ctx context.Context, mode storage.ModeGet, addr swarm.Addres
 	if err != nil {
 		if errors.Is(err, storage.ErrNotFound) {
 			// request from network
-			data, err := s.retrieval.RetrieveChunk(ctx, addr)
+			ch, err := s.retrieval.RetrieveChunk(ctx, addr, s.valid)
```

Comment:
I don't think this is a good idea after all. I know I came up with this line but this could introduce problems (e.g. it's easy to get deadlocks if there are locks added to net store at some points). Also we are verifying the chunk twice which is unnecessary work.

I do think it would be more clean to return the peer we received the chunk from in `RetrieveChunk` and then find a way to disconnect from that peer after the `!s.valid(ch)` check further below. This does of course mean that we first `Credit` the peer for the invalid chunk which might trigger settlement, but if that happens its only because we already received a lot of valid chunks before, so the loss is not that big.

Reply:
I have to disagree partly. 
The problem is, if we don't verify chunk in retrieve, we do not query more peers if we get an invalid chunk.
So the current change makes retries possible when an invalid chunk is returned.

Removed redundant validity check, passing validator functions should probably be benchmarked.

Reply:
this all depends on how we'd like to do accounting. if accounting is implicit and is part of the protocol abstraction like in the old swarm codebase then this has to be done as part of the retrieval protocol validation. i'm not sure if passing the actual validator function is the best solution here. it could be that we would be better off with some reusable validator component that can be passed around both to the protocols.

that being said, if accounting for a chunk would be done declaratively outside the protocol then this change is not needed and bumping the accounting counters can be done after the validator check originally on `L41` on the left. not sure what you have in mind tho. my personal opinion is that it is nicer that we have the validator check only in netstore, and not have these code duplications across the protocols. although if this is very much necessary to do in this manner it is certainly doable, and we might find some way to make it nice and clean.

Reply:
having thought this through, i think that it is no longer needed to have this validator on netstore at all of we go with this approach of inspecting whether the chunk is invalid of not on the protocol level. passing it along as a function reference is clunky and adds unnecessary indirection